elastic/elastic-agent-shipper
# Add infinite retry, fix monitoring

## What does this PR do?

Closes https://github.com/elastic/elastic-agent-shipper/issues/262

This PR adds infinite retry functionality, and completely refactors the elasticsearch health monitor. In detail

- If the `max_retries` is set to `-1`, this enables "infinite" retry, which just sets the ES MaxRetries to `MaxInt`.
- Because many common error conditions will now retry instead of trickling down to `OnFailure`, the monitor instead watches the raw HTTP return codes coming from the elasticsearch transport, and checks the deltas.
- The monitor will mark the ES output as unhealthy if zero successful HTTP codes are reported in a given period, and > 0 error conditions are reported across the reporting period

## Why is it important?

We want infinite retry.

## Checklist

- [ ] I have added an entry in `CHANGELOG.md` or `CHANGELOG-developer.md`.


## How to test this PR locally

- Pull down and build
- Run either standalone or with agent
- Set `max_retries` to `-1`:
```
outputs:
  default:
    log_level: debug
    type: elasticsearch
    hosts: [https://127.0.0.1:9200]
    username: "elastic"
    password: "changeme"
    bulk_max_size: 4096
    max_retries: -1
    ssl:
      verification_mode: none
    shipper:
      log_level: debug
      enabled: true
      bulk_max_size: 2048
```

- Ensure that events send properly
- Break Elasticsearch
- Run again, ensure that the shipper reports a degraded state
